test/test_device_spmv.cu

```diff
@@ -0,0 +1,598 @@
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED. IN NO EVENT SHALL NVIDIA CORPORATION BE LIABLE FOR ANY
+ * DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
+ * (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+ * LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
+ * ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+ * SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ *
+ ******************************************************************************/
+
+// Ensure printing of CUDA runtime errors to console
+#define CUB_STDERR
+
+#include <thrust/device_vector.h>
+#include <thrust/distance.h>
+#include <thrust/host_vector.h>
+#include <thrust/mismatch.h>
+#include <thrust/scan.h>
+
+#include <cub/device/device_spmv.cuh>
+#include <cub/util_debug.cuh>
+
+#include <iostream>
+#include <type_traits>
+#include <typeinfo>
+
+#include "test_util.h"
+
+bool g_verbose = false;
+
+//==============================================================================
+// Casts char types to int for numeric printing
+template <typename T>
+T print_cast(T val)
+{
+  return val;
+}
+
+int print_cast(char val) { return static_cast<int>(val); }
+
+int print_cast(signed char val) { return static_cast<int>(val); }
+
+int print_cast(unsigned char val) { return static_cast<int>(val); }
+
+//==============================================================================
+// Print a vector to out
+template <typename VectorT>
+void print_vector(std::ostream& out, const VectorT& vec)
+{
+  bool first = true;
+  for (const auto& val : vec)
+  {
+    if (!first)
+    {
+      out << ", ";
+    }
+    first = false;
+    out << print_cast(val);
+  }
+}
+
+//==============================================================================
+// Simple CSR matrix implementation.
+// HostStorage controls whether data is stored on the host or device.
+// Use the host_csr_matrix and device_csr_matrix aliases for code clarity.
+template <typename ValueT, bool HostStorage>
+struct csr_matrix
+{
+  csr_matrix(int num_rows, int num_cols)
+      : m_row_offsets(static_cast<size_t>(num_rows + 1), 0)
+      , m_num_rows(num_rows)
+      , m_num_columns(num_cols)
+  {}
+
+  // host/device conversion constructor
+  explicit csr_matrix(const csr_matrix<ValueT, !HostStorage>& other)
+      : m_values(other.m_values)
+      , m_row_offsets(other.m_row_offsets)
+      , m_column_indices(other.m_column_indices)
+      , m_num_rows(other.m_num_rows)
+      , m_num_columns(other.m_num_columns)
+      , m_num_nonzeros(other.m_num_nonzeros)
+  {}
+
+  // Note that this must append to the values array. Finish filling each row
+  // before adding to the next, and each row's columns must be added in order.
+  // Only non-zero elements should be added.
```

Comment:
This comment seems a bit outdated. We add zero elements now.

Reply:
Good catch! Updated.